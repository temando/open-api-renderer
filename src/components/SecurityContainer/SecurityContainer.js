import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Description from '../Description/Description'

import './SecurityContainer.scss'

export default class SecurityContainer extends PureComponent {
  render () {
    const { id, security } = this.props
    const { name, type, description } = security
    const isSimple = ['apiKey', 'http'].includes(security.type)

    return (
      <section className='security-container' id={id}>
        <h3>{name} <code className='scheme'>type={type}</code></h3>
        {description && <Description description={description} />}
        {isSimple && this.renderSimple(security)}
        {security.type === 'oauth2' && this.renderOAuth2(security)}
        {security.type === 'openIdConnect' && this.renderOpenIdConnect(security)}
      </section>
    )
  }

  renderSimple (security) {
    const { example, bearerFormat } = security
    let usage

    if (security.in === 'query') {
      usage = <p>To use this authentication scheme, add <code>{example}</code> to request URLs.</p>
    } else {
      usage = <p>To use this authentication scheme, pass <code>{example}</code>{bearerFormat ? `, formatted as ${bearerFormat}` : ''} as a request header.</p>
    }

    return (
      <div>
        {usage}
      </div>
    )
  }

  renderOAuth2 (security) {
    const { flows } = security

    return (
      <div>
        {Object.keys(flows).map((flowKey) => {
          const flow = flows[flowKey]

          return (
            <div className='flow-type' key={flowKey}>
              <h4><code>{flowKey}</code> flow</h4>
              <dl className='inline-pairs'>
                {flow.authorizationUrl && [
                  <dt key='auth'>Authorization URL</dt>,
                  <dd key='auth-value'>{flow.authorizationUrl}</dd>
                ]}
                {flow.tokenUrl && [
                  <dt key='token'>Token URL</dt>,
                  <dd key='token-value'>{flow.tokenUrl}</dd>
                ]}
                {flow.refreshUrl && [
                  <dt key='refresh'>Refresh URL</dt>,
                  <dd key='refresh-value'>{flow.refreshUrl}</dd>
                ]}
              </dl>
              <div className='scopes'>
                <h5>Available scopes</h5>
                {this.renderScopes(flow.scopes)}
              </div>
            </div>
          )
        })}
      </div>
    )
  }

  renderScopes (scopes) {
    return (
      <ul>
        {Object.keys(scopes).map(
          (scope) => <li key={scope}><strong>{scope}</strong> {scopes[scope]}</li>)
        }
      </ul>
    )
  }

  renderOpenIdConnect (security) {
    const { openIdConnectUrl } = security

    return (
      <div>
        <dl className='inline-pairs'>
          <dt>OpenID Connect URL</dt>
          <dd>{openIdConnectUrl}</dd>
        </dl>
      </div>
    )
  }
}

SecurityContainer.propTypes = {
  id: PropTypes.string,
  security: PropTypes.object
}
