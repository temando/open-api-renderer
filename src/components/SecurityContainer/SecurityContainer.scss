@import '../../base.scss';

.security-container {
  padding: 1rem 0;
  margin-bottom: 2rem;
  border-bottom: 1px solid $border-color;

  .scheme {
    display: inline-block;
    padding: .4rem;
    margin-left: .5rem;
    background-color: $navigation-background;
  }

  .flow-type {
    padding: 1rem 0;

    & + .flow-type {
      border-top: 1px dotted $border-color
    }
  }
}
