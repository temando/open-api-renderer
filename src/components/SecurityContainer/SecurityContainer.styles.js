export const styles = theme => ({
  'securityContainer': {
    padding: '1rem 0',
    borderBottom: `1px solid ${theme.borders.default}`
  },
  'scheme': {
    display: 'inline-block',
    marginLeft: '.5rem',
    backgroundColor: `${theme.backgrounds.nav}`
  },

  'flowType': {
    padding: '1rem 0',

    '& + &': {
      borderTop: `1px dotted ${theme.borders.default}`
    }
  },

  inlinePairs: {},
  scopes: {
    '& > li > span': {
      fontWeight: 600
    }
  }
})
