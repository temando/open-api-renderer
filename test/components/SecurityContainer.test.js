import React from 'react'
import SecurityContainer from './../../src/components/SecurityContainer/SecurityContainer'
import renderer from 'react-test-renderer'

describe('<SecurityContainer />', () => {
  it('renders apiKey type', () => {
    const security = {
      type: 'apiKey',
      name: 'api_key',
      in: 'header',
      example: 'api_key: credentials'
    }

    const tree = renderer.create(
      <SecurityContainer id='api_key' security={security} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders http type', () => {
    const security = {
      type: 'http',
      scheme: 'bearer',
      bearerFormat: 'JWT',
      name: 'JWT',
      example: 'Authorization: bearer credentials'
    }

    const tree = renderer.create(
      <SecurityContainer id='JWT' security={security} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders oauth2 type', () => {
    const security = {
      name: 'oauth2',
      type: 'oauth2',
      flows: {
        implicit: {
          authorizationUrl: 'https://example.com/api/oauth/dialog',
          scopes: {
            'write:pets': 'modify pets in your account',
            'read:pets': 'read your pets'
          }
        },
        authorizationCode: {
          authorizationUrl: 'https://example.com/api/oauth/dialog',
          tokenUrl: 'https://example.com/api/oauth/token',
          scopes: {
            'write:pets': 'modify pets in your account',
            'read:pets': 'read your pets'
          }
        }
      }
    }

    const tree = renderer.create(
      <SecurityContainer id='oauth2' security={security} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders openIdConnect type', () => {
    const security = {
      name: 'openIdConnect',
      type: 'openIdConnect',
      openIdConnect: 'https://example.com/openIdConnect'
    }

    const tree = renderer.create(
      <SecurityContainer id='openIdConnect' security={security} />
    )

    expect(tree).toMatchSnapshot()
  })
})
