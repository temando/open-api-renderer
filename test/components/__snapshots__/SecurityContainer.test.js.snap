// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SecurityContainer /> renders apiKey type 1`] = `
<section
  className="security-container"
  id="api_key"
>
  <h3>
    api_key
     
    <code
      className="scheme"
    >
      type=
      apiKey
    </code>
  </h3>
  <div>
    <p>
      To use this authentication scheme, pass 
      <code>
        api_key: credentials
      </code>
      
       as a request header.
    </p>
  </div>
</section>
`;

exports[`<SecurityContainer /> renders http type 1`] = `
<section
  className="security-container"
  id="JWT"
>
  <h3>
    JWT
     
    <code
      className="scheme"
    >
      type=
      http
    </code>
  </h3>
  <div>
    <p>
      To use this authentication scheme, pass 
      <code>
        Authorization: bearer credentials
      </code>
      , formatted as JWT
       as a request header.
    </p>
  </div>
</section>
`;

exports[`<SecurityContainer /> renders oauth2 type 1`] = `
<section
  className="security-container"
  id="oauth2"
>
  <h3>
    oauth2
     
    <code
      className="scheme"
    >
      type=
      oauth2
    </code>
  </h3>
  <div>
    <div
      className="flow-type"
    >
      <h4>
        <code>
          implicit
        </code>
         flow
      </h4>
      <dl
        className="inline-pairs"
      >
        <dt>
          Authorization URL
        </dt>
        <dd>
          https://example.com/api/oauth/dialog
        </dd>
      </dl>
      <div
        className="scopes"
      >
        <h5>
          Available scopes
        </h5>
        <ul>
          <li>
            <strong>
              write:pets
            </strong>
             
            modify pets in your account
          </li>
          <li>
            <strong>
              read:pets
            </strong>
             
            read your pets
          </li>
        </ul>
      </div>
    </div>
    <div
      className="flow-type"
    >
      <h4>
        <code>
          authorizationCode
        </code>
         flow
      </h4>
      <dl
        className="inline-pairs"
      >
        <dt>
          Authorization URL
        </dt>
        <dd>
          https://example.com/api/oauth/dialog
        </dd>
        <dt>
          Token URL
        </dt>
        <dd>
          https://example.com/api/oauth/token
        </dd>
      </dl>
      <div
        className="scopes"
      >
        <h5>
          Available scopes
        </h5>
        <ul>
          <li>
            <strong>
              write:pets
            </strong>
             
            modify pets in your account
          </li>
          <li>
            <strong>
              read:pets
            </strong>
             
            read your pets
          </li>
        </ul>
      </div>
    </div>
  </div>
</section>
`;

exports[`<SecurityContainer /> renders openIdConnect type 1`] = `
<section
  className="security-container"
  id="openIdConnect"
>
  <h3>
    openIdConnect
     
    <code
      className="scheme"
    >
      type=
      openIdConnect
    </code>
  </h3>
  <div>
    <dl
      className="inline-pairs"
    >
      <dt>
        OpenID Connect URL
      </dt>
      <dd />
    </dl>
  </div>
</section>
`;
